### PR TITLE
Faster type matching

### PR DESCRIPTION
**Pull request by laurit**
Nice matchers match based on the class that is being loaded, very nice matchers just match the class name.
Naughty matcher match based on the class hierarch (does this class extend X, does it implement Y).
Evil matchers match based on the complex criteria that involves other classes (does this class contain a method that overrides a method annotated with A in super class or interfaces). We have 4 such evil matchers:  [jax-rs 1](https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/main/instrumentation/jaxrs/jaxrs-1.0/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/jaxrs/v1_0/JaxrsAnnotationsInstrumentation.java), [jax-rs 2](https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/main/instrumentation/jaxrs/jaxrs-2.0/jaxrs-2.0-common/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/jaxrs/v2_0/JaxrsAnnotationsInstrumentation.java), [jax-ws](https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/main/instrumentation/jaxws/jaxws-jws-api-1.1/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/jaxws/jws/v1_1/JwsAnnotationsInstrumentation.java) and [external annotations](https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/main/instrumentation/external-annotations/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/extannotations/ExternalAnnotationInstrumentation.java). The last one can hopefully be fixed by removing `hasSuperType` (I can't see a reason why it is there in the first place), the rest should be disable by default, deleted or written in some other way.
The problem with naughty matchers is that they need to find bytes for super types, parse them and cache them to avoid repeating this. If there are many classes this can be quite costly. This PR introduces an alternative strategy. When a class is loaded, first ensure its super types are loaded. During matching find super type with `ClassLoder.findLoadedClass` and use the already loaded class for navigating type hierarchy. If just the name or the modifiers of the super type are needed then there is no need to find the actual bytes for the class. Additionally using this strategy fixes a shortcoming of current strategy. Based on just  the class loader of class being loaded it is not always possible to find bytes for all of its super types as in OSGI environments super type of super type need not be visible to current class. There should be less stack traces for `net.bytebuddy.pool.TypePool$Resolution$NoSuchTypeException` in smoke tests for server thats use OSGI.

### CODE REVIEW

**Comment by mateuszrzeszutek** [comment]
> We have 4 such evil matchers: [jax-rs 1](https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/main/instrumentation/jaxrs/jaxrs-1.0/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/jaxrs/v1_0/JaxrsAnnotationsInstrumentation.java), [jax-rs 2](https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/main/instrumentation/jaxrs/jaxrs-2.0/jaxrs-2.0-common/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/jaxrs/v2_0/JaxrsAnnotationsInstrumentation.java), [jax-ws](https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/main/instrumentation/jaxws/jaxws-jws-api-1.1/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/jaxws/jws/v1_1/JwsAnnotationsInstrumentation.java) and [external annotations](https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/main/instrumentation/external-annotations/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/extannotations/ExternalAnnotationInstrumentation.java). The last one can hopefully be fixed by removing `hasSuperType` (I can't see a reason why it is there in the first place), the rest should be disable by default, deleted or written in some other way.

I think we could probably disable them by default and explicitly state in our docs that they're very expensive to apply, if you use JAX-RS then you have to enable it manually.

**Comment by laurit** [comment]
> I think we could probably disable them by default and explicitly state in our docs that they're very expensive to apply, if you use JAX-RS then you have to enable it manually.

These expensive instrumentations are only used to create a span for the annotated method. We already have other spans for both JAX-RS and JAX-WS create by framework specific instrumentations that are probably more useful than the expensive one.

**Comment by trask** [comment]
💯

**Comment by anuraaga** [comment]
If the jax instrumentations are missing classloader matchers, can we add them and still keep the instrumentation enabled by default?

**Comment by laurit** [comment]
> If the jax instrumentations are missing classloader matchers, can we add them and still keep the instrumentation enabled by default?

They have the class loader matchers. The problem is that these apis are present in a lot application servers which will enable these instrumentation even when they aren't really used. I have enable them for now, we can deal with this later.

**Review by trask — APPROVED** [review]
❤️❤️❤️

> These expensive instrumentations are only used to create a span for the annotated method. We already have other spans for both JAX-RS and JAX-WS create by framework specific instrumentations that are probably more useful than the expensive one.

I'd be interested in seeing the impact to the tests of disabling these expensive ones in a follow-up. I think it would depend on what exactly we lose by having these off-by-default, but I'm very open to it, as startup performance is a big deal.

File: muzzle/src/main/java/io/opentelemetry/javaagent/tooling/muzzle/AgentCachingPoolStrategy.java [hunk 1]
```diff
@@ -241,89 +289,264 @@ public void clear() {
...
+            List<TypeDescription> result = new ArrayList<>();
+            for (Generic interfaceDescription : interfaces) {
+              String interfaceName = interfaceDescription.getTypeName();
+              Class<?> interfaceClass = findLoadedClass(classLoader, interfaceName);
+              if (interfaceClass != null) {
+                // using raw type
+                result.add(newTypeDescription(interfaceClass));
+              } else {
+                result.add(interfaceDescription.asErasure());
+              }
```
[the remark this answers is not in the document]
Comment (laurit) [reply muzzle/src/main/java/io/opentelemetry/javaagent/tooling/muzzle/AgentCachingPoolStrategy.java:448]: This method returns a generic type https://github.com/raphw/byte-buddy/blob/fbd78e1868c53b0c73b9528f821465bb3173794c/byte-buddy-dep/src/main/java/net/bytebuddy/description/type/TypeDescription.java#L459-L469 but I opted to just use a raw type. My understanding is that you can think of this as the difference between `getSuperClass` and `getGenericSuperclass`. You can't construct correct value for `getGenericSuperclass` when you only have the result of  `getSuperClass`. For our purposes using the raw type (the equivalent of `getSuperClass`) should suffice, I can't imagine a situation where we would need this info during matching. All of this seems complicated, looking at
https://github.com/raphw/byte-buddy/blob/fbd78e1868c53b0c73b9528f821465bb3173794c/byte-buddy-dep/src/main/java/net/bytebuddy/description/type/TypeDescription.java#L8685-L8691 we'd need to use some kind of projection to combine the child class with the interface to implement this.

File: muzzle/src/main/java/io/opentelemetry/javaagent/tooling/muzzle/AgentCachingPoolStrategy.java [hunk 2]
```diff
@@ -241,89 +289,264 @@ public void clear() {
...
+    /**
+     * Non-lazily resolves a type name.
+     *
+     * @param name The name of the type to resolve.
+     * @return The resolution for the type of this name.
+     */
+    protected TypePool.Resolution doResolve(String name) {
+      TypePool.Resolution resolution = cacheProvider.find(name);
+      if (resolution == null) {
+        resolution = cacheProvider.register(name, super.doDescribe(name));
```
[the remark this answers is not in the document]
Comment (laurit) [reply muzzle/src/main/java/io/opentelemetry/javaagent/tooling/muzzle/AgentCachingPoolStrategy.java:324]: This is copied from https://github.com/raphw/byte-buddy/blob/fbd78e1868c53b0c73b9528f821465bb3173794c/byte-buddy-dep/src/main/java/net/bytebuddy/pool/TypePool.java#L921
The `doDescribe` in this class returns a lazy resolution, this method is called to actually locate and parse the class bytes

**Comment by laurit** [comment]
> I'd be interested in seeing the impact to the tests of disabling these expensive ones in a follow-up. I think it would depend on what exactly we lose by having these off-by-default, but I'm very open to it, as startup performance is a big deal.

I did some rough measurements starting liferay. I measured invocation count & time spent for 2 point
- locating the class. I wrapped the locator in https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/77dae8d1deb4705dfa99b829d9671b8f6f99ddaa/muzzle/src/main/java/io/opentelemetry/javaagent/tooling/muzzle/AgentLocationStrategy.java#L36
- type matching. I wrapped the matcher from type instrumentation in https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/77dae8d1deb4705dfa99b829d9671b8f6f99ddaa/javaagent-tooling/src/main/java/io/opentelemetry/javaagent/tooling/instrumentation/InstrumentationModuleInstaller.java#L105

This pr.
```
match count=5409259, time=5247ms
locate count=10559, time=3333ms
```

Class loader optimization matcher disabled, this ensures that the jax-rs/ws annotation matchers have something to match.
```
match count=10154036, time=16974ms
locate count=32589, time=9313ms
```

Class loader optimization matcher disabled, jax-rs/ws annotation matchers disabled.
```
match count=9965431, time=6738ms
locate count=9279, time=2094ms
```

Class loader optimization matcher disabled, jax-rs/ws annotation matchers disabled, matchers that match based on annotation (external annotations, withspan, spring-ws) disabled.
```
match count=9368871, time=4852ms
locate count=3258, time=1008ms
```
Byte-buddy locates bytes for all annotations on all methods if you have a matcher that matches based on annotation. I hope I can convince byte-buddy author to change this or give us some way to just match based on the name without locating the annotation type.

In retrospect I realized that I probably should have counted only the classes located during matching. Then the last test would give
```
locate count=45, time=19ms
```
which looks good.

File: muzzle/src/main/java/io/opentelemetry/javaagent/tooling/muzzle/AgentCachingPoolStrategy.java [hunk 3]
```diff
@@ -241,89 +289,268 @@ public void clear() {
...
+      public boolean isResolved() {
+        return doResolve(name).isResolved();
+      }
+
+      private volatile TypeDescription cached;
+
+      @Override
+      public TypeDescription resolve() {
+        // unlike byte-buddy implementation we cache the descriptor to avoid having to find
+        // super class and interfaces multiple times
```
Comment (mateuszrzeszutek) [inline muzzle/src/main/java/io/opentelemetry/javaagent/tooling/muzzle/AgentCachingPoolStrategy.java:351]: 👍

**Review by mateuszrzeszutek — APPROVED** [review]

File: muzzle/src/main/java/io/opentelemetry/javaagent/tooling/muzzle/AgentTooling.java [hunk 4]
```diff
@@ -22,9 +24,18 @@ public static AgentLocationStrategy locationStrategy(ClassLoader bootstrapProxy)
...
+
+  public static void init(ClassLoader classLoader) {
+    bootstrapProxy = classLoader;
   }
 
   private AgentTooling() {}
+
+  private static class PoolStrategyHolder {
+    private static final AgentBuilder.PoolStrategy POOL_STRATEGY =
+        new AgentCachingPoolStrategy(locationStrategy(bootstrapProxy));
```
[the remark this answers is not in the document]
Comment (laurit) [reply muzzle/src/main/java/io/opentelemetry/javaagent/tooling/muzzle/AgentTooling.java:39]: This is also used from muzzle. WDYT if I add an SPI for getting the bootstrap proxy?

**Comment by trask** [comment]
> jax-rs/ws annotation matchers disabled

wow, those matchers are slow!! (esp. now that the inheritance hierarchy matchers have been optimized in this PR 👍)

**Review by trask — APPROVED** [review]

**Review by anuraaga — APPROVED** [review]

File: muzzle/src/main/java/io/opentelemetry/javaagent/tooling/muzzle/AgentTooling.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (mateuszrzeszutek) [reply muzzle/src/main/java/io/opentelemetry/javaagent/tooling/muzzle/AgentTooling.java]: That'd probably solve the problem, but wouldn't just creating a bootstrap proxy CL in the muzzle module be a bit simpler?

[the remark this answers is not in the document]
Comment (laurit) [reply muzzle/src/main/java/io/opentelemetry/javaagent/tooling/muzzle/AgentTooling.java]: It depends on whether muzzle code needs the same bootstrap proxy as agent (as agent jar is appended to bootclasspath getting access to resources in agent needs a bit of work). Having multiple pool strategy instances could complicate this a bit as then they wouldn't share the cache unless some stuff is made static.

**Comment by trask** [comment]
🎉

**Pull request merged**